# Module 3 Code Challenge

## Overview

This assessment is designed to test your understanding of Module 3 material. It covers:

* Gradient Descent
* Logistic Regression
* Classification Metrics
* Decision Trees

_Read the instructions carefully._ You will be asked both to write code and respond to a few short answer questions.

### Note on the short answer questions

For the short answer questions, _please use your own words._ The expectation is that you have **not** copied and pasted from an external source, even if you consult another source to help craft your response. While the short answer questions are not necessarily being assessed on grammatical correctness or sentence structure, do your best to communicate yourself clearly.

---
## Part 1: Gradient Descent [Suggested Time: 20 min]
---

![best fit line](visuals/best_fit_line.png)

The best fit line that goes through the scatterplot up above can be generalized in the following equation: $$y = mx + b$$

Of all the possible lines, we can prove why that particular line was chosen using the plot down below:

![](visuals/cost_curve.png)

where RSS is defined as the residual sum of squares:

$$ 
\begin{align}
RSS &= \sum_{i=1}^n(actual - expected)^2 \\
&= \sum_{i=1}^n(y_i - \hat{y})^2 \\
&= \sum_{i=1}^n(y_i - (mx_i + b))^2
\end{align}
$$ 

### 1.1) What is a more generalized name for the RSS curve above? How could a machine learning model use this curve?

"""
The curve above is a cost function curve. This plots the loss measurement on the y-axis and the parameter you are optimizing for on the x axis. Machine learning models use this with gradient descent to determine the value, in this case slope, that gives gives the lowest loss value. Using gradient descent, a point on the parabola is chosen. Using the derivative to calculate the slope, this, along with your step size, determines how the parameters are changed which moves the point further down the slope to a better loss value.
"""

### 1.2) Would you rather choose a $m$ value of 0.08 or 0.05 from the RSS curve up above? Explain your reasoning.

"""
An $m$ value of 0.05 would be better because it corresponds with a lower RSS value approaching the bottom of the curve. This means that there was less difference between our actual and predicted values. 
"""

![](visuals/gd.png)

### 1.3) Using the gradient descent visual from above, explain why the distance between estimates in each step is getting smaller as more steps occur with gradient descent.

"""
The distance between estimates in each step decreases as more steps occur in order to end up at the bottom of the graph which corresponds to the lowest loss value (in this case RSS). If the distances did not decrease they would overshoot the best value for the RSS.
"""

### 1.4) What does the learning rate do in the gradient descent algorithm? Explain how a very small and a very large learning rate would affect the gradient descent.

"""
The gradient descent algorithm chooses the step size according to the slope and learning rate. As the points move further down the line the slope approaches 0 and minimizes the step sizes to hopefully converge at, and not overshoot, the optimal point at the bottom of the parabola. A learning rate that is too large will overshoot the optimal value while a learning rate that is too small with either take too long to reach the bottom or will get stuck at a saddle point.
"""

---
## Part 2: Logistic Regression [Suggested Time: 15 min]
---

### 2.1) Why is logistic regression typically better than linear regession for modeling a binary target/outcome?

"""
The regression line in linear regression would not capture the binary target/outcome because a binary target has to be thought of as a probability plotted from 0 to 1. For example if you had many people write down a number on a piece of paper and you try to determine whether a number is a 5 you only have "yes it is a 5" or "no it is not a 5". Logistic regression uses the sigmoid function to transform the regression line into an "S" shape to better capture this relationship. 
"""

### 2.2) What is one advantage that logistic regression can have over other classification methods?

"""
Logistic regression tends to be more interpretable than other models. Other classification methods like decision trees have too many complex interactions to directly interpret what a particular feature's effect on the target variable is.
"""

---
## Part 3: Classification Metrics [Suggested Time: 20 min]
---

![cnf matrix](visuals/cnf_matrix.png)

### 3.1) Using the confusion matrix above, calculate precision, recall, and F-1 score.

Show your work, not just your final numeric answer

In [3]:
true_positives = 30
false_positives = 12
true_negatives = 54
false_negatives = 4

In [4]:
# Your code here to calculate precision
precision = true_positives/(true_positives + false_positives)
precision

0.7142857142857143

In [5]:
# Your code here to calculate recall
recall = true_positives/(true_positives + false_negatives)
recall

0.8823529411764706

In [6]:
# Your code here to calculate F-1 score
f1_score = 2*((precision * recall)/(precision + recall))
f1_score

0.7894736842105262

<img src = "visuals/many_roc.png" width = "700">

### 3.2) Which ROC curve from the above graph is the best? Explain your reasoning. 

Note: each ROC curve represents one model, each labeled with the feature(s) inside each model.

"""
The best model is the "All Features" model. This is because it gives us the highest true positive rate (guessing true and getting true) while minimizing the false positive rate (guessing true and getting false). It also seems to have the highest AUC which is a metric for the quality of our models' predictions.
"""

### Logistic Regression Example

The following cell includes code to train and evaluate a model

In [7]:
# Run this cell without changes

# Include relevant imports
import pickle
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler 
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_curve, roc_auc_score

network_df = pickle.load(open('write_data/sample_network_data.pkl', 'rb'))

# partion features and target 
X = network_df.drop('Purchased', axis=1)
y = network_df['Purchased']

# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2019)

# scale features
scale = StandardScaler()
scale.fit(X_train)
X_train = scale.transform(X_train)
X_test = scale.transform(X_test)

# build classifier
model = LogisticRegression(C=1e5, solver='lbfgs')
model.fit(X_train, y_train)
y_test_pred = model.predict(X_test)

# get the accuracy score
print(f'The classifier has an accuracy score of {round(accuracy_score(y_test, y_test_pred), 3)}.')

The classifier has an accuracy score of 0.956.


### 3.3) Explain how the distribution of `y` shown below could explain the very high accuracy score.

In [8]:
# Run this cell without changes

y.value_counts()

0    257
1     13
Name: Purchased, dtype: int64

"""
The distribution of y is an example of class imbalance. The majority class of "0" has 257 values while the minority class of "1" only has 13. If the model "stupidly" predicts 0 every time the imbalance in the majority class will lead to a high accuracy score.
"""

### 3.4) What method could you use to address the issue discovered in Question 3.3? 

"""
For this example a metric like F1 would be a better choice. You could also address the class imbalance by using up/downsampling techniques like SMOTE or Tomek Links to even out the classes. 
"""

---
## Part 4: Decision Trees [Suggested Time: 20 min]
---

### Concepts 
You're given a dataset of **30** elements, 15 of which belong to a positive class (denoted by *`+`* ) and 15 of which do not (denoted by `-`). These elements are described by two attributes, A and B, that can each have either one of two values, true or false. 

The diagrams below show the result of splitting the dataset by attribute: the diagram on the left hand side shows that if we split by attribute A there are 13 items of the positive class and 2 of the negative class in one branch and 2 of the positive and 13 of the negative in the other branch. The right hand side shows that if we split the data by attribute B there are 8 items of the positive class and 7 of the negative class in one branch and 7 of the positive and 8 of the negative in the other branch.

<img src="visuals/decision_stump.png">

### 4.1) Which one of the two attributes resulted in the best split of the original data? How do you select the best attribute to split a tree at each node? 

It may be helpful to discuss splitting criteria.

"""
Attribute A resulted in the best split of the original data. If you calculate the gini impurity score, "A" gets the lower score and therefore results in the best split. You can also see that a resulted in more homogeneous nodes.
"""

### Decision Tree Example

In this section, you will use decision trees to fit a classification model to the wine dataset. The data is the results of a chemical analysis of wines grown in the same region in Italy by three different cultivators. There are thirteen different measurements taken for different constituents found in the three types of wine.

In [10]:
# Run this cell without changes

# Relevant imports 
import pandas as pd 
import numpy as np 
from sklearn.datasets import load_wine

# Load the data 
wine = load_wine()
X, y = load_wine(return_X_y=True)
X = pd.DataFrame(X, columns=wine.feature_names)
y = pd.Series(y)
y.name = 'target'
df = pd.concat([X, y.to_frame()], axis=1)
df.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  target  
0                          3.92   1065.0       0  
1                          3.40   1050.0       0  
2                          3.17   1185.0       0  
3                          3.45   1480.0       0  
4                          2.93    735.0       0

In [11]:
# Run this cell without changes
# Get the shape of the DataFrame 
df.shape

(178, 14)

In [12]:
# Run this cell without changes
# Get the distribution of the target variable 
y.value_counts()

1    71
0    59
2    48
Name: target, dtype: int64

### 4.2) Split the data into training and test sets. Create training and test sets with `test_size=0.5` and `random_state=1`.

In [13]:
# Replace None with appropriate code  
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=23)

### 4.3) Fit a decision tree model with scikit-learn to the training data. Use parameter defaults, except for `random_state=1`. Use the fitted classifier to generate predictions for the test data.

You can use the Scikit-learn DecisionTreeClassifier (docs [here](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html))

In [16]:
# Your code here 
from sklearn.tree import DecisionTreeClassifier
dt_clf = DecisionTreeClassifier(random_state=1)
dt_clf_fit = dt_clf.fit(X_train,y_train)
y_pred = dt_clf_fit.predict(X_test)

### 4.4) Obtain the accuracy score of the predictions on the test set. 

You can use the `sklearn.metrics` module.

In [20]:
# Your code imports here
from sklearn.metrics import accuracy_score
# Replace None with appropriate code 

print('Accuracy Score:', accuracy_score(y_test, y_pred))

Accuracy Score: 0.9777777777777777


### 4.5) Produce a confusion matrix for the predictions on the test set. 

You can use the `sklearn.metrics` module.

In [22]:
# Your code imports here
from sklearn.metrics import confusion_matrix
# Your code here 
cm = confusion_matrix(y_test, y_pred)
cm

array([[15,  0,  0],
       [ 0, 20,  0],
       [ 0,  1,  9]])

### 4.6) Do the accuracy score or confusion matrix reveal any substantial problems with this model's performance? Explain your answer.

"""
There are issues with the model's performance. The accuracy score appears to be too high meaning that the model could be overfitting. The same idea is true in the confusion matrix. Because we used a decision tree classifier it is possible that it created too many branches and narrowed down the nodes too far resulting in it learning each individual point as opposed to the general relationship in the data. 